src/core/exec-invoke.c

```diff
@@ -4877,6 +5001,44 @@ int exec_invoke(
                 }
         }
 
+        /* Unshare a new PID namespace before setting up mounts to ensure /proc/ is mounted with only processes in PID namespace visible.
+         * Note PrivatePIDs=yes implies MountAPIVFS=yes so we'll always ensure procfs is remounted. */
+        if (needs_sandboxing && exec_needs_pid_namespace(context)) {
+                if (params->pidref_transport_fd < 0) {
+                        *exit_status = EXIT_NAMESPACE;
+                        return log_exec_error_errno(context, params, r, "PidRef socket is not set up: %m");
+                }
+
+                /* If we had CAP_SYS_ADMIN prior to joining the user namespace, then we are privileged and don't need
+                 * to check if we can mount /proc/.
+                 *
+                 * We need to check prior to entering the user namespace because if we're running unprivileged or in a
+                 * system without CAP_SYS_ADMIN, then we can have CAP_SYS_ADMIN in the current user namespace but not
+                 * once we unshare a mount namespace. */
+                r = has_cap_sys_admin ? 1 : can_mount_proc(context, params);
+                if (r > 0) {
+                        r = setup_private_pids(context, params);
+                        if (r < 0) {
+                                *exit_status = EXIT_NAMESPACE;
+                                return log_exec_error_errno(context, params, r, "Failed to set up pid namespace: %m");
+                        }
+
+                        if (r > 0)
+                                /* Return value > 0 means we are in the parent process. Thus, return immediately in order to exit ASAP.
+                                * The rest of the setup will continue in the child process. */
+                                return 0;
```

Comment:
so, I think this should actually be an `_exit(EXIT_SUCCESS)`, as discussed above.

Reply:
Done! Had to add some logic to send errno from parent -> child via pipe2. Its similar to can_mount_proc() but we can't have the child wait for the parent so we always have to send the errno.

Confirmed this fixes the ASAN issue, thanks!